### PR TITLE
[WIP] Bugfix for "--topics/--services cannot work without extra option before bag file path"

### LINKED ISSUE #1628: `ros2 bag play --topics/--services` cannot work without extra option before bag file path
## Description

`--topics` and `--services` option arguments cannot work without having the extra option before the `bag file path`.

### Expected Behavior

it should work even without extra option before bag file path.

```bash
root@tomoyafujita:~/ros2_ws/colcon_ws# ros2 bag play --topics chatter rosbag2_2024_04_25-10_55_16/
[INFO] [1714067979.777474156] [rosbag2_player]: Set rate to 1
[INFO] [1714067979.783672709] [rosbag2_player]: Adding keyboard callbacks.
[INFO] [1714067979.783702132] [rosbag2_player]: Press SPACE for Pause/Resume
[INFO] [1714067979.783707171] [rosbag2_player]: Press CURSOR_RIGHT for Play Next Message
[INFO] [1714067979.783712411] [rosbag2_player]: Press CURSOR_UP for Increase Rate 10%
[INFO] [1714067979.783716795] [rosbag2_player]: Press CURSOR_DOWN for Decrease Rate 10%
[INFO] [1714067979.783814715] [rosbag2_player]: Playback until timestamp: -1
```

### Actual Behavior

```bash
root@tomoyafujita:~/ros2_ws/colcon_ws# ros2 bag play --topics chatter rosbag2_2024_04_25-10_55_16/
usage: ros2 bag play [-h] [-s {sqlite3,my_test_plugin,mcap,my_read_only_test_plugin}]
                     [--read-ahead-queue-size READ_AHEAD_QUEUE_SIZE] [-r RATE] [--topics topic [topic ...]]
                     [--services service [service ...]] [-e REGEX] [-x EXCLUDE_REGEX] [--exclude-topics topic [topic ...]]
                     [--exclude-services service [service ...]] [--qos-profile-overrides-path QOS_PROFILE_OVERRIDES_PATH]
                     [-l] [--remap REMAP [REMAP ...]] [--storage-config-file STORAGE_CONFIG_FILE] [--clock [CLOCK] |
                     --clock-topics CLOCK_TOPICS [CLOCK_TOPICS ...] | --clock-topics-all] [-d DELAY]
                     [--playback-duration PLAYBACK_DURATION]
                     [--playback-until-sec PLAYBACK_UNTIL_SEC | --playback-until-nsec PLAYBACK_UNTIL_NSEC]
                     [--disable-keyboard-controls] [-p] [--start-offset START_OFFSET] [--wait-for-all-acked TIMEOUT]
                     [--disable-loan-message] [--publish-service-requests]
                     [--service-requests-source {service_introspection,client_introspection}]
                     bag_path
ros2 bag play: error: the following arguments are required: bag_path
```

## To Reproduce

1. start some publishers.

```bash
ros2 topic pub -r 1 /greetings  example_interfaces/msg/String "{data: 'Hello fron greeting topic'}"
ros2 topic pub -r 1 /chatter std_msgs/msg/String "{data: 'Hello on greeting topic'}"
```

2. record all topics.

```bash
root@tomoyafujita:~/ros2_ws/colcon_ws# ros2 bag record -a
[INFO] [1714067717.014419278] [rosbag2_recorder]: Press SPACE for pausing/resuming
[INFO] [1714067717.018212708] [rosbag2_recorder]: Listening for topics...
[INFO] [1714067717.018257318] [rosbag2_recorder]: Event publisher thread: Starting
[INFO] [1714067717.026120002] [rosbag2_recorder]: Subscribed to topic '/rosout'
[INFO] [1714067717.029323308] [rosbag2_recorder]: Subscribed to topic '/parameter_events'
[INFO] [1714067717.031138205] [rosbag2_recorder]: Subscribed to topic '/events/write_split'
[INFO] [1714067717.031189731] [rosbag2_recorder]: Recording...
[INFO] [1714067721.022214861] [rosbag2_recorder]: Subscribed to topic '/greetings'
[INFO] [1714067724.435570665] [rosbag2_recorder]: Subscribed to topic '/chatter'
[INFO] [1714067736.480997299] [rosbag2_recorder]: Pausing recording.
[INFO] [1714067736.481258938] [rosbag2_cpp]: Writing remaining messages from cache to the bag. It may take a while
[INFO] [1714067736.488452363] [rosbag2_recorder]: Event publisher thread: Exiting
[INFO] [1714067736.488516934] [rosbag2_recorder]: Recording stopped
[INFO] [1714067736.577997951] [rclcpp]: signal_handler(signum=2)
```

3. try to replay the topic from the bag file with `--topics` optional argument.

```bash
root@tomoyafujita:~/ros2_ws/colcon_ws# ros2 bag play --topics chatter greetings rosbag2_2024_04_25-10_55_16/
usage: ros2 bag play [-h] [-s {sqlite3,my_test_plugin,mcap,my_read_only_test_plugin}]
                     [--read-ahead-queue-size READ_AHEAD_QUEUE_SIZE] [-r RATE] [--topics topic [topic ...]]
                     [--services service [service ...]] [-e REGEX] [-x EXCLUDE_REGEX] [--exclude-topics topic [topic ...]]
                     [--exclude-services service [service ...]] [--qos-profile-overrides-path QOS_PROFILE_OVERRIDES_PATH]
                     [-l] [--remap REMAP [REMAP ...]] [--storage-config-file STORAGE_CONFIG_FILE] [--clock [CLOCK] |
                     --clock-topics CLOCK_TOPICS [CLOCK_TOPICS ...] | --clock-topics-all] [-d DELAY]
                     [--playback-duration PLAYBACK_DURATION]
                     [--playback-until-sec PLAYBACK_UNTIL_SEC | --playback-until-nsec PLAYBACK_UNTIL_NSEC]
                     [--disable-keyboard-controls] [-p] [--start-offset START_OFFSET] [--wait-for-all-acked TIMEOUT]
                     [--disable-loan-message] [--publish-service-requests]
                     [--service-requests-source {service_introspection,client_introspection}]
                     bag_path
ros2 bag play: error: the following arguments are required: bag_path
```

## System (please complete the following information)

 - OS: Ubuntu 24.04 Nobel
 - ROS 2 Distro: Rolling
 - Install Method: source code
 - Version: https://github.com/ros2/ros2/commit/4d07f58591920794676261a14c2ba9d0a7bf2746

## Additional context

as a work-around, if we have an extra option before the bag file path, it works as expected.

the following example command includes `-s mcap` before bag file path.

```bash
root@tomoyafujita:~/ros2_ws/colcon_ws# ros2 bag play --topics chatter greetings -s mcap rosbag2_2024_04_25-10_55_16/
[INFO] [1714067979.777474156] [rosbag2_player]: Set rate to 1
[INFO] [1714067979.783672709] [rosbag2_player]: Adding keyboard callbacks.
[INFO] [1714067979.783702132] [rosbag2_player]: Press SPACE for Pause/Resume
[INFO] [1714067979.783707171] [rosbag2_player]: Press CURSOR_RIGHT for Play Next Message
[INFO] [1714067979.783712411] [rosbag2_player]: Press CURSOR_UP for Increase Rate 10%
[INFO] [1714067979.783716795] [rosbag2_player]: Press CURSOR_DOWN for Decrease Rate 10%
[INFO] [1714067979.783814715] [rosbag2_player]: Playback until timestamp: -1
```

### PR DESCRIPTION
**Pull request by MichaelOrlov**
- Closes https://github.com/ros2/rosbag2/issues/1628
- Add a quote guarded space-separated list type to player CLI
- For example, after the fix, the list of topics could be passed as:
1. `ros2 bag play --topics "topic1 topic2" bagpath`
2. `ros2 bag play --topics 'topic1 topic2' bagpath`
3. `ros2 bag play --topics topic1 bagpath`

### CODE REVIEW

File: ros2bag/ros2bag/verb/play.py [hunk 1]
```diff
@@ -37,131 +37,135 @@ def positive_float(arg: str) -> float:
...
+        '--read-ahead-queue-size', type=int, default=1000,
+        help='size of message queue rosbag tries to hold in memory to help deterministic '
+             'playback. Larger size will result in larger memory needs but might prevent '
+             'delay of message playback.')
+    parser.add_argument(
+        '-r', '--rate', type=check_positive_float, default=1.0,
+        help='rate at which to play back messages. Valid range > 0.0.')
+    parser.add_argument(
+        '--topics', type=lambda s: list(s.split(' ')), default=[], metavar='topic',
+        help='Space-delimited list of topics to play.')
```
Comment (fujitatomoya) [inline ros2bag/ros2bag/verb/play.py]: - `ros2 bag play --topics chatter rosbag2_2024_04_25-10_55_16` works.
- `ros2 bag play --topics 'chatter greetings' rosbag2_2024_04_25-10_55_16` works.
- `ros2 bag play --topics "chatter greetings" rosbag2_2024_04_25-10_55_16` works.

but,

- `ros2 bag play --topics chatter greetings rosbag2_2024_04_25-10_55_16` fails with `ros2 bag play: error: argument bag_path: Bag path 'greetings' does not exist!`

i think we need to add, the list needs to be quoted.

```suggestion
    parser.add_argument(
        '--topics', type=lambda s: list(s.split(' ')), default=[], metavar='topic',
        help='Space-delimited and quoted list of topics to play.')
```

File: ros2bag/ros2bag/verb/play.py [hunk 2]
```diff
@@ -37,131 +37,135 @@ def positive_float(arg: str) -> float:
...
+    parser.add_argument(
+        '--services', type=lambda s: list(s.split(' ')), default=[], metavar='service',
+        help='Space-delimited list of services to play.')
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/play.py]: As fujitatomoya mentioned, "and quoted" should also be added to the description.

File: ros2bag/ros2bag/verb/play.py [hunk 3]
```diff
@@ -37,131 +37,135 @@ def positive_float(arg: str) -> float:
...
+        help='Path to a yaml file defining storage specific configurations. '
+             'See storage plugin documentation for the format of this file.')
+    clock_args_group = parser.add_mutually_exclusive_group()
+    clock_args_group.add_argument(
+        '--clock', type=positive_float, nargs='?', const=40, default=0,
+        help='Publish to /clock at a specific frequency in Hz, to act as a ROS Time Source. '
+             'Value must be positive. Defaults to not publishing.')
+    clock_args_group.add_argument(
+        '--clock-topics', type=lambda s: list(s.split(' ')), default=[],
+        help='List of topics separated by spaces that will trigger a /clock update '
```
Comment (Barry-Xu-2018) [inline ros2bag/ros2bag/verb/play.py]: As fujitatomoya mentioned, "and quoted" should also be added to the description.

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018  @fujitatomoya Thank you for the preliminary review and opinion.   
I agree that we need to add in the parameters help section that the list needs to be surrounded by quotes.
I will follow up, fix it, and add more tests. I am also going to change lists for the recorder to be consistent.

**Comment by Barry-Xu-2018** [comment]
@MichaelOrlov 

>  I am also going to change lists for the recorder to be consistent.

If you have higher-priority tasks, I can take care of this modification.

**Comment by MichaelOrlov** [comment]
@Barry-Xu-2018 @fujitatomoya Almost done. 
I added changes for recorder and burst verbs to be consistent and fixed integration unit tests, which started failing after the changes.  

What I don't like in the current implementation is that in the recorder the `topics` CLI argument is positional.  
For instance, if one wants to specify a list of services and list of topics for recording the CLI command should be
```bash
ros2 bag record --services "service1 service2" "topic1 topic2" --output path/to/bag
```
It works currently, but the sequence of the parameters `--services "service1 service2" "topic1 topic2"` could cause confusion and could be a source of mistakes because the **topics** parameter can't be specified explicitly. i.e. `--topics`.
Please note that we inherited this behaviour from the past when we were implementing the services recording feature.

I am going to make the `topics` CLI argument as non-positional argument to be specified explicitly as `--topics`.  
Thoughts or concerns about it?

**Comment by Barry-Xu-2018** [comment]
@MichaelOrlov 

Yes, the argument for topics in the record command are styled differently from those in the play command. I considered making topics argument of the record command similar to the play command while I modified the code of record.  Considering compatibility, I ultimately decided not to propose this idea.  

I strongly agree with changing topics to a non-positional argument.

**Comment by fujitatomoya** [comment]
> It works currently, but the sequence of the parameters --services "service1 service2" "topic1 topic2" could cause confusion and could be a source of mistakes because the topics parameter can't be specified explicitly. i.e. --topics.

agree, i do not think this is good behavior at all.

> I am going to make the topics CLI argument as non-positional argument to be specified explicitly as --topics.
> Thoughts or concerns about it?

sounds good to me.

**Comment by MichaelOrlov** [comment]
- Closing as will not be merged in favor of the existent workaround with trailing `--`.
It turns out that the Python argparse class has a builtin option to split positional arguments by `--`
It is not well documented, but exists https://stackoverflow.com/questions/26985650/argparse-do-not-catch-positional-arguments-with-nargs/26986546#26986546

**Comment by fujitatomoya** [comment]
@MichaelOrlov thanks, good to know.

**Pull request closed without merging**